### DISCUSSION TITLE
Question regarding stack size before each instruction

### DISCUSSION THREAD

**Discussion by m-carrasco** [jbevain/cecil#756] [Q&A] [answered]
Hi there,

I need to compute the stack size at each CIL instruction while assuming it is legal bytecode. To do it, I'm reading how the CodeWriter computes the max stack size of a method. Having read it, I wonder about [this behavior](https://github.com/jbevain/cecil/blob/a72c1b1e11762d940f95ffb662dbdf7f0de92a28/Mono.Cecil.Cil/CodeWriter.cs#L409):

`branch_stack_size = System.Math.Max (branch_stack_size, computed_size); `

My understanding is that if the bytecode were legal, then the above calculation would be redundant. In the sense that every instruction that may jump to the current one should have the same stack size.  I am not implying that there is a bug in the implementation, I guess Cecil tries to aid the user to generate a legal output even if the input is illegal. 

Do you think that I am comprehending how things work? Am I missing a case? 

Kind regards,
Manuel.

**Comment by ltrzesniewski** [comment, marked as answer]
> I guess Cecil tries to aid the user to generate a legal output even if the input is illegal.

I believe so. Here's what ECMA-335 says in I.12.3.2.1:

> The type state of the stack (the stack depth and types of each element on
> the stack) at any given point in a program shall be identical for all possible control flow paths.
> For example, a program that loops an unknown number of times and pushes a new element on
> the stack at each iteration would be prohibited. 

If you're interested, I wrote a simple stack depth checker based on Cecil's code [here](https://github.com/ltrzesniewski/InlineIL.Fody/blob/master/src/InlineIL.Tests/Debug/StackStateTests.cs). It basically computes the stack size for each instruction, and fails if there's a mismatch. It's not pretty (I just wanted to check if another part of my code was based on correct assumptions), but it verifies code from all loaded assemblies at that point, and here are the results:

> .NET 5.0.6
> Invalid methods: 0 / 65154 in 59 assemblies

> .NET Framework 4.8.4341.0
> Invalid methods: 0 / 56785 in 18 assemblies

**Reply by m-carrasco** [reply]
Thanks for your answer and for sharing your code. I'll take it into account as well as cecil's one.
